### DISCUSSION TITLE
Process modes

### DISCUSSION THREAD

**Discussion by GauBen** [Hexagon/pup#32] [Ideas]
The idea of this feature is to offer alternate definitions of processes, that make them mutually exclusive:

```jsonc
{
  "processes": [
    {
      "id": "frontend",
      "cmd": "yarn workspace frontend run start",
      "modes": { // I'm not sure if I prefer an array or an object here
        "dev": {
          "cmd": "yarn workspace frontend run dev"
        }
      }
    }
  ]
}
```

Here's how things would work:
- The default mode would be referred as `default`
- Starting processes under certain modes could be done as `pup start frontend@dev api` (start frontend with dev mode, api with default mode), or `pup start --mode dev frontend api` (start frontend and api as dev mode, fallback to default mode)
- It would not be possible to start `frontend@default` and `frontend@dev` at the same time, the cli could offer a prompt to switch mode if another is already running

I haven't thought this through, but I feel it would be a valuable addition.

Use cases:
- Start processes in dev (hot reload) or classic mod
- Start mocked processes instead of real ones

```jsonc
{
  "processes": [
    {
      "id": "bin-thing",
      "dependsOn": ["database", "broker", "api", ...],
      "modes": {
        "mocked": {
          "dependsOn": [] // Nothing needed to make it run
        }
      }
    }
  ]
}
```

Also, I feel like `processes` should be a record `{ processName: processDeclaration }`, this would make `id` required and unique by design

**Comment by Hexagon** [comment]
This also sounds great, everything in the "mode" would simply override the default settings of the process recursively. Simple, and usable! Every process could have a target mode set by pup start id@mode, if watchdog find a difference in target mode and current mode, it would immediately kill the current process and switch mode? Or should it switch at next restart?

Are you interested in implementing this?

**Reply by GauBen** [reply]
> it would immediately kill the current process and switch mode? Or should it switch at next restart?

The choice is yours! I'd implement it as such:

- If `pup start id@mode` was run in a tty, show a prompt offering to replace the process
- If a `--mode-mismatch-behavior [replace|ignore]` flag is present, do as the flag say
- Otherwise, fail with a specific status code
